internal/controllerlib/option.go

```diff
@@ -113,6 +113,10 @@ func WithInformer(getter InformerGetter, filter Filter, opt InformerOption) Opti
 				}
 			},
 		})
+		if err != nil {
+			// Shouldn't really happen.
+			panic(die(fmt.Sprintf("got error from AddEventHandler: %s", err.Error())))
```

Comment:
This seems scary. It feels more backward-similar to ignore the returned `err` instead of panic.

Reply:
Yes, this was a bit of a hack to get the PR to compile and run in CI. I wanted to discuss this with you and @benjaminapetersen.

In all version of Kube prior to 1.26, [the function in question had no return value](https://github.com/kubernetes/kubernetes/blob/v1.25.6/staging/src/k8s.io/client-go/tools/cache/shared_informer.go#L135-L138) and therefore could not error.

In 1.26, now it has two return values, one of which is an error. Looking at the implementation of that library function, it [can only error for one reason](https://github.com/kubernetes/kubernetes/blob/6b55f097bb2140381a58312aeede37fc76a0762e/staging/src/k8s.io/client-go/tools/cache/shared_informer.go#L583), which shouldn't happen in the way that we are using it. However, this could always change in future versions and it could start returning errors for other situations. We use this helper with all of our controllers, so we need to know if it starts returning an error after some future dependency bump. If it did actually return an error, this could only happen during app startup, so we should catch it right away in CI.

Usually, a nicer approach to error handling would be to let the error bubble up through the callers, but in this case the error is inside `WithInformer()` which is trying to provide the functional option pattern, so it can't easily return an error without significantly reworking the controllerlib to make all Options return errors, and also to make the `Run`, `Start`, and `Prepare` functions handle these errors. This is possible but would would be a significant change, especially considering that these functions make heavy use of background goroutines, so it is much simpler if Options cannot error.

The controllerlib also has other similar panics which should not happen in practice already, such as [this one](https://github.com/vmware-tanzu/pinniped/blob/7ff3b3d9cbbde7a245f3078c1ed6af6f67001ad7/internal/controllerlib/option.go#L152), probably for the same reasons that I mentioned above.

After thinking through it more, I think using a panic here is a reasonable solution. What do y'all think?

Reply:
I agree we should acknowledge the `err` (instead of swallowing it and continuing operations).